url43a0865bb0_kevinjdolan_darwin_tgz-pJ8-tui_DriverJ8/src/darwInvest/data/index/TickerIndex.java

```diff
@@ -0,0 +1,107 @@
+package darwInvest.data.index;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.Map;
+import java.util.TreeMap;
+import darwInvest.data.Ticker;
+
+/**
+ * The TickerIndex contains an index of tickers.
+ * 
+ * @author Andrew Perrault, Kevin Dolan
+ */
+public class TickerIndex implements Serializable {
+
+	private static final long serialVersionUID = 693612329281611960L;
+	private String source;
+	private Map<String, IndexItem> index;
+	
+	/**
+	 * Instantiate a new blank ticker index
+	 * @param source the name of the data source
+	 */
+	public TickerIndex(String source) {
+		this.source = source;
+		index = new TreeMap<String, IndexItem>();
+	}
+
+	/**
+	 * Add a ticker to this index, replace it if it
+	 * already exists
+	 * @param ticker the ticker to add
+	 */
+	public void addTicker(Ticker ticker) {
+		IndexItem item = new IndexItem(ticker);
+		index.put(ticker.getSymbol(), item);
+	}
+
+	/**
+	 * Determine whether a symbol exists in the index
+	 * @param symbol the symbol to check for
+	 * @return		 true if it exists
+	 */
+	public boolean symbolExists(String symbol) {
+		return index.get(symbol) == null;
+	}
+	
+	/**
+	 * Return a ticker object pointed to by the index
+	 * @param symbol the symbol to load
+	 * @return		 the unserialized ticker object
+	 */
+	public Ticker getTickerFile(String symbol) {
+		File fileName = new File("StockData/" + source + "/" + symbol + ".stock");
+		
+		Ticker ticker = null;
+		FileInputStream fis = null;
+		ObjectInputStream in = null;
+		try {
+			fis = new FileInputStream(fileName);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *fis*. The resource is referenced at lines 65, 66. The resource is closed at line 67. There are other execution paths that do not contain closure statements, for example, when ObjectInputStream.readObject throws an exception. Ensure *fis* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).